hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3774,18 +3784,28 @@ private TableGroup createTableGroup(TableGroup parentTableGroup, SqmPath<?> join
 				querySpec.getFromClause().addRoot( tableGroup );
 			}
 			else {
+				final SqlAstJoinType sqlAstJoinType;
+				if ( currentClauseStack.getCurrent() != Clause.SELECT
+						&& getInferredValueMapping() == null
```

Comment:
The reason to use `LEFT` join only when the inferred mapping is null is we need it just for nullness predicates, otherwise we always want `INNER` joins for i.e. comparison operations.

WDYT of something like:  
```
// If we have an inferred value mapping we are using this path "against" another, 
// and we always want INNER Join semantics in that case
```

Reply:
The inferred mapping can be null in the following potentially problematic cases:

1. top-level select item
2. group by expression
3. tuple expression
4. select item of exists subquery

Case 1. and 4. should be fixed by changing `prepareForSelection` like I suggested above. Then you also wouldn't need to care about the SELECT clause in here. No idea how to fix 2. and 3. though. Problematic means that e.g. `from MyEntity e group by e.notFoundAssociation` would produce a `left join` for the association, just like a `from MyEntity e where (e.id,e.notFoundAssociation)=(1,:param)` would.

Reply:
I have implemented a solution which passes an `allowLeftJoins` flag only when visiting an `is null` predicate. This allows searching for an existing join of the correct type (`left` for not found to-one associations) and eventually upgrade it if the context requires `inner` joins.